antlr_ast/ast.py

```diff
@@ -283,3 +175,560 @@ def reportContextSensitivity(
+                kwargs["helper"] = self.helper
+            return transform_function(node, **kwargs)
+
+        return transformer_method
+
+
+# TODO: test: if 'visit' in method, it has to be as 'visit_'
+class BaseNodeTransformer(NodeTransformer):
+    def __init__(self, registry: BaseNodeRegistry):
+        self.helper = TransformerHelper(registry)
+
+    def visit(self, node):
+        # TODO: I think transform_  + node.__class__.__name__ would be better/clearer then
+        #  as the node methods don't need to do any visiting (which is completely done by visit and generic_visit)
+        method = "visit_" + node.__class__.__name__
+        transformer = getattr(self, method, None)
+
+        if transformer is None:
+            return self.generic_visit(node)
+        else:
+            alias = transformer(node)
+            if isinstance(alias, AliasNode) or alias == node:
+                # this prevents infinite recursion and visiting
+                # AliasNodes with a name that is also the name of a BaseNode
+                self.generic_visit(alias)
+            else:
+                # visit BaseNode (e.g. result of Transformer method)
+                if isinstance(alias, list):
+                    # Transformer method can return array instead of node
+                    alias = [self.visit(el) for el in alias]  # TODO: test
+                elif isinstance(alias, BaseNode):
+                    alias = self.visit(alias)
+
+            return alias
+
+    def visit_Terminal(self, terminal):
+        return self.terminal_visitor(terminal)
+
+    def visit_str(self, string):
+        return self.terminal_visitor(string)
+
+    def terminal_visitor(self, terminal):
+        """Helper to transparently handle Terminal"""
+        return terminal
+
+    @classmethod
+    def bind_alias_nodes(cls, alias_classes):
+        for item in alias_classes:
+            if getattr(item, "_rules", None) is not None:
+                item.bind_to_transformer(cls)
+
+
+class TransformerHelper:
+    def __init__(self, registry: BaseNodeRegistry):
+        self.registry = registry
+
+    def isinstance(self, *args):
+        return self.registry.isinstance(*args)
+
+
+def get_alias_nodes(items):
+    return list(
+        filter(
+            lambda item: inspect.isclass(item) and issubclass(item, AliasNode), items
+        )
+    )
+
+
+def simplify_tree(tree, unpack_lists=True, in_list=False):
+    """Recursively unpack single-item lists and objects where fields and labels only reference a single child
+
+    :param tree: the tree to simplify (mutating!)
+    :param unpack_lists: whether single-item lists should be replaced by that item
+    :param in_list: this is used to prevent unpacking a node in a list as AST visit can't handle nested lists
+    """
+    # TODO: copy (or (de)serialize)? outside this function?
+    if isinstance(tree, BaseNode) and not isinstance(tree, Terminal):
```

Comment:
The way I understand this is it collapses single-child nodes into 1 edge. Can you send me an example SQL statement to see this in action? I tested with `SELECT * FROM 'test'`.

Also, would be nice to have a test for this.